### PR TITLE
Adding ability for SQS source CRD to use annotations

### LINKED ISSUE #986: Please support Kube2IAM auth for SQS
I would like to use the credentials built into the POD via Kube2IAM instead of cred file

### PR DESCRIPTION
**Pull request by AceHack**
Fixes #986 

https://github.com/knative/eventing-contrib/issues/986

## Proposed Changes

Added the ability for AwsSqsSource CRD to expose pod annotations, this will allow for using Kube2IAM credentials.  The existing credentials file is also supported.  This is a backward-compatible change.

- 🎁 Add new feature

**Release Note**

```release-note
Added the ability for AwsSqsSource CRD to expose pod annotations, this will allow for using Kube2IAM credentials.  The existing credentials file is also supported.
```

**Docs**

awssqs/samples/README.md was updated as part of this PR

### CODE REVIEW

File: awssqs/samples/README.md [hunk 1]
```diff
@@ -55,6 +57,14 @@ Then create a secret for the downloaded key:
 kubectl -n knative-sources create secret generic awssqs-source-credentials --from-file=credentials=PATH_TO_CREDENTIALS_FILE
 ```
 
+### Set up [Kube2IAM](https://github.com/jtblin/kube2iam) credentials
+
+NOTE: If not using Kube2IAM skip this step
+
+Replace the `AWS_IAM_ROLE` place holders in `samples/awssqs-source-kube2iam.yaml`.
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
Replace the `AWS_IAM_ROLE` place holders in
`samples/awssqs-source-kube2iam.yaml`.
```

File: awssqs/samples/README.md [hunk 2]
```diff
@@ -55,6 +57,14 @@ Then create a secret for the downloaded key:
...
+
+`AWS_IAM_ROLE` should be replaced with your AWS IAM role that looks something like `AWS_IAM_ROLE`
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
`AWS_IAM_ROLE` should be replaced with your AWS IAM role that looks something
like `AWS_IAM_ROLE`
```

File: awssqs/samples/README.md [hunk 3]
```diff
@@ -66,7 +76,7 @@ ko apply -f awssqs/config/
 Note that if the `Source` Service Account secret is in a non-default location,
 you will need to update the YAML first.
 
-Replace the place holders in `samples/awssqs-source.yaml`.
+Replace the place holders in `samples/awssqs-source.yaml` or `samples/awssqs-source-kube2iam.yaml` depending on credential type used.
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
Replace the place holders in `samples/awssqs-source.yaml` or
`samples/awssqs-source-kube2iam.yaml` depending on credential type used.
```

File: awssqs/samples/README.md [hunk 4]
```diff
@@ -75,11 +85,19 @@ Replace the place holders in `samples/awssqs-source.yaml`.
   sed -i "s|QUEUE_URL|$QUEUE_URL|" awssqs-source.yaml
   ```
 
-Now deploy `awssqs-source.yaml`.
+Now deploy `awssqs-source.yaml` or `awssqs-source-kube2iam.yaml` depending on credential type used.
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
Now deploy `awssqs-source.yaml` or `awssqs-source-kube2iam.yaml` depending on
credential type used.
```

File: awssqs/samples/README.md [hunk 5]
```diff
@@ -75,11 +85,19 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
 
-```shell
-ko apply -f awssqs/samples/awssqs-source.yaml
-```
+- Credentials File
+
+   ```shell
+   ko apply -f awssqs/samples/awssqs-source.yaml
+   ```
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source.yaml
  ```
```

File: awssqs/samples/README.md [hunk 6]
```diff
@@ -75,11 +85,19 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
+
+- Kube2IAM credentials
+
+   ```shell
+   ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
+   ```
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
  ```
```

File: awssqs/samples/README.md [hunk 7]
```diff
@@ -55,6 +57,15 @@ Then create a secret for the downloaded key:
...
 ```
 
+### Set up [Kube2IAM](https://github.com/jtblin/kube2iam) credentials
+
+NOTE: If not using Kube2IAM skip this step
+
+Replace the `AWS_IAM_ROLE` place holders in
+`samples/awssqs-source-kube2iam.yaml`.
+
+`AWS_IAM_ROLE` should be replaced with your AWS IAM role that looks something like `AWS_IAM_ROLE`
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
`AWS_IAM_ROLE` should be replaced with your AWS IAM role that looks something
like `AWS_IAM_ROLE`
```

File: awssqs/samples/README.md [hunk 8]
```diff
@@ -66,7 +77,7 @@ ko apply -f awssqs/config/
 Note that if the `Source` Service Account secret is in a non-default location,
 you will need to update the YAML first.
 
-Replace the place holders in `samples/awssqs-source.yaml`.
+Replace the place holders in `samples/awssqs-source.yaml` or `samples/awssqs-source-kube2iam.yaml` depending on credential type used.
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
Replace the place holders in `samples/awssqs-source.yaml` or
`samples/awssqs-source-kube2iam.yaml` depending on credential type used.
```

File: awssqs/samples/README.md [hunk 9]
```diff
@@ -75,11 +86,19 @@ Replace the place holders in `samples/awssqs-source.yaml`.
   sed -i "s|QUEUE_URL|$QUEUE_URL|" awssqs-source.yaml
   ```
 
-Now deploy `awssqs-source.yaml`.
+Now deploy `awssqs-source.yaml` or `awssqs-source-kube2iam.yaml` depending on credential type used.
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
Now deploy `awssqs-source.yaml` or `awssqs-source-kube2iam.yaml` depending on
credential type used.
```

File: awssqs/samples/README.md [hunk 10]
```diff
@@ -75,11 +86,19 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
 
-```shell
-ko apply -f awssqs/samples/awssqs-source.yaml
-```
+- Credentials File
+
+   ```shell
+   ko apply -f awssqs/samples/awssqs-source.yaml
+   ```
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source.yaml
  ```
```

File: awssqs/samples/README.md [hunk 11]
```diff
@@ -75,11 +86,19 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
+
+- Kube2IAM credentials
+
+   ```shell
+   ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
+   ```
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
  ```
```

File: awssqs/samples/README.md (hunk 8 above)
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
Replace the place holders in `samples/awssqs-source.yaml` or
`samples/awssqs-source-kube2iam.yaml` depending on credential type used.
```

File: awssqs/samples/README.md (hunk 9 above)
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
Now deploy `awssqs-source.yaml` or `awssqs-source-kube2iam.yaml` depending on
credential type used.
```

File: awssqs/samples/README.md (hunk 10 above)
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source.yaml
  ```
```

File: awssqs/samples/README.md (hunk 11 above)
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
  ```
```

File: awssqs/samples/README.md [hunk 12]
```diff
@@ -75,11 +87,20 @@ Replace the place holders in `samples/awssqs-source.yaml`.
   sed -i "s|QUEUE_URL|$QUEUE_URL|" awssqs-source.yaml
   ```
 
-Now deploy `awssqs-source.yaml`.
+Now deploy `awssqs-source.yaml` or `awssqs-source-kube2iam.yaml` depending on
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
Now deploy `awssqs-source.yaml` or `awssqs-source-kube2iam.yaml` depending on
credential type used.
```

File: awssqs/samples/README.md [hunk 13]
```diff
@@ -75,11 +87,20 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
+credential type used.
 
-```shell
-ko apply -f awssqs/samples/awssqs-source.yaml
-```
+- Credentials File
+
+   ```shell
+   ko apply -f awssqs/samples/awssqs-source.yaml
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source.yaml
  ```
```

File: awssqs/samples/README.md [hunk 14]
```diff
@@ -75,11 +87,20 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
+   ```
+
+- Kube2IAM credentials
+
+   ```shell
+   ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
  ```
```

File: awssqs/samples/README.md [hunk 15]
```diff
@@ -75,11 +87,20 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
+credential type used.
 
-```shell
-ko apply -f awssqs/samples/awssqs-source.yaml
-```
+- Credentials File
+
+   ```shell
+   ko apply -f awssqs/samples/awssqs-source.yaml
+   ```
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source.yaml
  ```
```

File: awssqs/samples/README.md [hunk 16]
```diff
@@ -75,11 +87,20 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
+
+- Kube2IAM credentials
+
+   ```shell
+   ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
+   ```
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
  ```
```

File: awssqs/samples/README.md [hunk 17]
```diff
@@ -75,11 +87,18 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
+credential type used.
 
-```shell
-ko apply -f awssqs/samples/awssqs-source.yaml
-```
+- Credentials File
+
+  ```shell
+  ko apply -f awssqs/samples/awssqs-source.yaml
+
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion

  ```

```

File: awssqs/samples/README.md [hunk 18]
```diff
@@ -75,11 +87,18 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
+- Kube2IAM credentials
+
+  ```shell
+  ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ```shell
  ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
  ```
```

File: awssqs/samples/README.md (hunk 17 above)
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion

  ```

```

File: awssqs/samples/README.md [hunk 19]
```diff
@@ -75,11 +87,18 @@ Replace the place holders in `samples/awssqs-source.yaml`.
...
```
Comment (mattmoor) [inline awssqs/samples/README.md]: Format markdown:
```suggestion
  ko apply -f awssqs/samples/awssqs-source-kube2iam.yaml
  ```
```

File: awssqs/samples/README.md (hunk 17 above)
Comment (AceHack) [reply awssqs/samples/README.md]: I don't understand this suggestion

File: awssqs/samples/README.md (hunk 19 above)
Comment (AceHack) [reply awssqs/samples/README.md]: I don't understand this suggestion

File: awssqs/cmd/receive_adapter/main.go [hunk 20]
```diff
@@ -49,6 +49,15 @@ func getRequiredEnv(envKey string) string {
 	return val
 }
 
+func getEnv(envKey string) string {
+	val, defined := os.LookupEnv(envKey)
+	if !defined {
+		return ""
+	}
+
+	return val
```
Comment (mattmoor) [inline awssqs/cmd/receive_adapter/main.go]: os.Getenv?

Comment (AceHack) [reply awssqs/cmd/receive_adapter/main.go]: Yep, I made that change, thank you.

**Comment by n3wscott** [comment]
Cool! Looks simple enough!

/lgtm
/approve

**Pull request merged**